<a href="https://colab.research.google.com/github/pedroregato/Alura-MachineLearning/blob/main/Aula02.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**02
Lendo dados da internet e manipulando os mesmos**

Neste aula, trabalharemos como uma outra [fonte de dados](https://gist.githubusercontent.com/guilhermesilveira/2d2efa37d66b6c84a722ea627a897ced/raw/10968b997d885cbded1c92938c7a9912ba41c615/tracking.csv) para nosso projeto.

Nela, encontramos uma tabela com quatro colunas diferentes, e na qual cada item representa um usuário que acessou um determinado site. Neste site, temos três páginas diferentes: a principal (home); a how it works, que explica o funcionamento do site; e contact, para entrar em contato.

Em cada uma dessas três colunas, que são as features, temos 1 ou 0, indicando se um usuário específico acessou essa página ou não, respectivamente. Por enquanto não nos importa saber a ordem de páginas acessada pelo usuário.

Também temos uma quarta coluna (bought) indicando se esse usuário comprou o produto (1) ou não (0). Abaixo temos uma parte da tabela para demonstração:

home	how_it_works	contact	bought
1	1	0	0
1	1	0	0
1	1	0	0
1	1	0	0
1	0	1	1
1	1	0	0
Clicando no botão raw, podemos acessar o formato padrão do arquivo, que apresenta o seguinte aspecto:

In [ ]:
home,how_it_works,contact,bought
1,1,0,0
1,1,0,0
1,1,0,0
1,1,0,0
1,1,1,1
1,1,0,0

Nesse formato os elementos são separados por vírgulas, cada linha representa um usuário e cada coluna uma das features ou a informação de que o usuário comprou ou não um produto.

De volta ao Google Colab, selecionaremos "File > NewPython 3 notebook" para criar uma nova área de trabalho. Criado o novo arquivo, o chamaremos de Introdução a Machine Learning Classificação - 2.ipynb.

Nós queremos ler o arquvo cru da tabela anterior (.csv). Para isso, usaremos a biblioteca Pandas. A maneira padrão de importar essa biblioteca é import pandas as pd. Em seguida, com pd.read_csv(uri), faremos a leitura do arquivo .csv. Você pode copiar essa uri do código abaixo:

In [2]:
import pandas as pd

uri = "https://gist.githubusercontent.com/guilhermesilveira/2d2efa37d66b6c84a722ea627a897ced/raw/10968b997d885cbded1c92938c7a9912ba41c615/tracking.csv"
pd.read_csv(uri)

home  how_it_works  contact  bought
0      1             1        0       0
1      1             1        0       0
2      1             1        0       0
3      1             1        0       0
4      1             1        0       0
..   ...           ...      ...     ...
94     0             0        1       0
95     1             1        1       1
96     0             0        1       0
97     0             1        0       0
98     0             0        0       0

[99 rows x 4 columns]

Ao pressionarmos "Shift + Enter", será exibida a tabela com os dados do arquivo cru. Neste caso, temos todos os dados disponíveis neste único arquivo, portanto usaremos a variável genérica dados para representar essas informações. Nosso objetivo inicial é imprimir somente as primeiras linhas da tabela, o que é feito com head():

In [3]:
import pandas as pd

uri = "https://gist.githubusercontent.com/guilhermesilveira/2d2efa37d66b6c84a722ea627a897ced/raw/10968b997d885cbded1c92938c7a9912ba41c615/tracking.csv"
dados = pd.read_csv(uri)
dados.head()

home  how_it_works  contact  bought
0     1             1        0       0
1     1             1        0       0
2     1             1        0       0
3     1             1        0       0
4     1             1        0       0

Contudo, ao refletirmos sobre o modelo que estamos adotando, seria mais interessante que a coluna bought fosse separada das outras três, afinal ela corresponde no modelo da função ao y*, enquanto nossas features (home, *howitworks e contact) representam **x.

Para selecionarmos especificamente as três primeiras colunas, usaremos a sintaxe dados[["home","how_it_works","contact"]]. Como são várias colunas, precisamos passar esses conteúdos dentro de um array. Ao executarmos a código, teremos a tabela atualizada:


home	how_it_works	contact
1	1	0
1	1	0
1	1	0
1	1	0
1	0	1
No código, chamaremos essas features de x. Para y, atriburemos dados[["bought"]], ou seja, somente a coluna bought:


In [5]:
x= dados[["home","how_it_works","contact"]]
y= dados[["bought"]]

No caso de y, não é necessário utilizar o colchete extra (ou seja, [[]]), pois estamos extraindo apenas uma única coluna. Podemos imprimir somente a coluna "bought" utilizando print(y) ou y.head() (para exibir apenas os cinco primeiros itens dessa lista):


Name: bought, dtype: int64
0
0
0
0
0
Algumas vezes preferimos trabalhar com as features em português. Para modificar as nomeações em inglês que já estamos utilizando, usaremos uma funcionalidade do Pandas que possibilita a renomeação das colunas: dados.rename(columns). Em seguida, passaremos mapa, um dicionário do Python em que declaramos qual será o nome das respectivas colunas:

In [6]:
mapa = { 
    "home" : "principal",
    "how_it_works" : "como_funciona",
    "contact" : "contato",
    "bought" : "comprou"
}
dados.rename(columns = mapa)

principal  como_funciona  contato  comprou
0           1              1        0        0
1           1              1        0        0
2           1              1        0        0
3           1              1        0        0
4           1              1        0        0
..        ...            ...      ...      ...
94          0              0        1        0
95          1              1        1        1
96          0              0        1        0
97          0              1        0        0
98          0              0        0        0

[99 rows x 4 columns]

Essa renomeção deve dialogar com as necessidades do seu projeto, já que talvez abreviações ou mesmo os nomes em inglês sejam mais interessantes dependendo de cada caso. Essa célula de código deve vir antes da atribuição de x e y. Além disso, essa alteração não altera a tabela original, portanto precisaremos atribuir dados.rename() para a variável dados:

In [8]:
mapa = { 
    "home" : "principal",
    "how_it_works" : "como_funciona",
    "contact" : "contato",
    "bought" : "comprou"
}
dados = dados.rename(columns = mapa)

Agora se executarmos novamente o código, veremos uma mensagem de erro:

KeyError: "['home', 'how_it_works' e 'contact'] not in index"

Isso porque os termos em inglês, que correspondiam às colunas, não existem mais em nosso projeto. No momento em que carregamos o arquivo, a primeira ação a ser tomada é renomear os itens necessários para que o projeto seja executado:

In [9]:
x= dados[["principal","como_funciona","contato"]]
y= dados[["comprou"]]

x.head()

principal  como_funciona  contato
0          1              1        0
1          1              1        0
2          1              1        0
3          1              1        0
4          1              1        0

Com x.head(), imprimiremos os primeiros cinco elementos da tabela para garantirmos que o código continua funcionando:


principal	como_funciona	contato
1	1	0
1	1	0
1	1	0
1	1	0
Podemos fazer o mesmo procedimento com y.head():

Name: comprou, dtype: int64
0
0
0
0
0
Agora que temos x e y separados, devemos nos questionar: devemos treinar os algorítimos com todos os dados? Se fizermos isso, não teremos o que testar, pois dessa forma a máquina já terá as respostas corretas previamente, sendo incapaz de prever dados fora desse conjunto. De alguma maneira, precisamos sempre separar os dados de treino e os de teste para não enviesarmos os resultados produzidos.

Primeiramente, com dados.shape verificaremos quantos elementos temos nos dados e o formato deles:


In [10]:
dados.shape

(99, 4)

Veremos que em nosso arquivo há 99 linhas e 4 colunas. Separaremos em média 25% para testar o algorítimo, e o restante (cerca de 75% dos dados) para o treinamento. Portanto, para treino_x, coletaremos os primeiros 75 elementos (treino_x = x[:75]). Podemos utilizar treino_x.shape para verificar se o número de elementos está de fato correto:

In [11]:
treino_x = x[:75]
treino_x.shape

(75, 3)

Como resultado, teremos 75 elementos e 3 colunas. Da mesma forma, treino_y deve receber y[:75]. Já teste_x e teste_y deverão se receber os dados a partir do elemento 75, ou seja, x[75:] e y[75:].

Para confirmarmos se as matrizes estão com a quantidade correta de elementos 24), acionaremos teste_y.shape. Por fim, registraremos essas informações imprimindo o tamanho (len()) de treino_x e treino_y utilizando print().

In [13]:
treino_x = x[:75]
treino_y = y[:75]
teste_x = x[75:]
teste_y = y[75:]
teste_y.shape

print("Treinaremos com %d elementos e testaremos com %d elementos" % (len(treino_x), len(teste_x)))

Treinaremos com 75 elementos e testaremos com 24 elementos


Para treinarmos e executarmos o algorítimo usaremos a mesma motodologia do exemplo de porcos e cachorros; do sklearn.svm importaremos LinearSVC, e treinaremos o modelo com os dados treino_x e treino_y.

In [14]:
from sklearn.svm import LinearSVC

modelo = LinearSVC()
modelo.fit(treino_x, treino_y)

/usr/local/lib/python3.7/dist-packages/sklearn/utils/validation.py:760: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


LinearSVC(C=1.0, class_weight=None, dual=True, fit_intercept=True,
          intercept_scaling=1, loss='squared_hinge', max_iter=1000,
          multi_class='ovr', penalty='l2', random_state=None, tol=0.0001,
          verbose=0)

Feito o treino, podemos fazer as previsoes baseadas no teste_x, passando modelo.predict(). Em seguida, compararemos as previsões com teste_y. Para isso, usaremos o accuracy_score() (sem nos esquecermos de importar esse método de sklearn.metrics), passando como argumentos teste_y e previsoes.

Esse método deve nos retornar a acurácia, ou seja, a taxa de acertos do modelo. Neste momento, já podemos formatar o resultado como porcentagem e imprimi-lo na tela:

In [15]:
from sklearn.svm import LinearSVC
from sklearn.metrics import accuracy_score

modelo = LinearSVC()
modelo.fit(treino_x, treino_y)
previsoes = modelo.predict(teste_x)

acuracia = accuracy_score(teste_y, previsoes) * 100
print("A acurácia foi %.2f%%" % acuracia)

A acurácia foi 95.83%


/usr/local/lib/python3.7/dist-packages/sklearn/utils/validation.py:760: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


Na tela, teremos:

A acurácia foi 95.83%

Por meio desse exemplo, tivemos um contato mais realista com tratamentos de algorítimos de machine learning, uma vez que importamos a fonte de dados e realizamos os tratamentos necessários, renomeando colunas e separando os dados de treino e teste até que finalmente executamos o modelo. Ainda há elementos que podem ser melhorados, e é justamente isso que estudaremos a seguir.